### DISCUSSION TITLE
is_val/has_val

### DISCUSSION THREAD

**Discussion by Beliar83** [biojppm/rapidyaml#228] [Q&A] [answered]
Heya,
in what context should one use is_val and when has_val?
Is `node.has_val() || node.is_val()` okay for situations I just want to know if the node has something that can be converted with either `atox` or `from_chars`? (I don't know the target type at that point, so I have to try to convert to determine that)

**Comment by biojppm** [comment, marked as answer]
To clarify:
* `has_val()` if it has a val regardless of having a key, so it is true both for sequence and map members
* `is_val()` if it has a val and does NOT have a key, so it is true only for sequence members

Here's the relevant code:
![image](https://user-images.githubusercontent.com/325344/158577184-ad2497a6-38f8-47b3-a88f-5b43b9d189c2.png)

Eg:
```yaml
seq:
  - val0
  - val1
map:
  key0: val0
  key1: val1
```
Then the following holds:
```c++
assert( ! tree["seq"].has_val()); // it's a container, so it does not have a val
assert( ! tree["map"].has_val()); // it's a container, so it does not have a val

assert(   tree["seq"][0].has_val());
assert(   tree["seq"][1].has_val());
assert(   tree["seq"][0].is_val());
assert(   tree["seq"][1].is_val());

assert(   tree["map"][0].has_val());
assert(   tree["map"][1].has_val());
assert( ! tree["map"][0].is_val()); // map member, so is_val is false because the node has a key
assert( ! tree["map"][1].is_val()); // map member, so is_val is false because the node has a key
```

**Reply by Beliar83** [reply]
Okay, thanks, I think I understand now.
